kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -1189,7 +1203,18 @@ public CompletableFuture<Long> recoverTxEntries(
             Executor executor) {
         if (!kafkaConfig.isKafkaTransactionCoordinatorEnabled()) {
             // no need to scan the topic, because transactions are disabled
-            return CompletableFuture.completedFuture(Long.valueOf(0));
+            return CompletableFuture.completedFuture(0L);
+        }
+        if (!isBrokerIndexMetadataInterceptorConfigured(persistentTopic.getBrokerService())) {
```

Comment:
If we don't have this check, the `UpgradeTest` will encounter an NPE problem, but we can't disable the Kafka transaction in that test right now, because newly Kafka client requires the producer id. @Demogorgon314